_posts/2019-01-13-dask-cudf-first-steps.md

```diff
@@ -0,0 +1,437 @@
+
+gdf = dask_cudf.read_csv('data/nyc/many/*.csv')  # wrap around many CSV files
+
+>>> gdf.passenger_count.sum().compute()
+184464740
+```
+
+*Also note, NYC Taxi ridership is significantly less than it was a few years ago*
+
+
+What I'm excited about in the example above
+-------------------------------------------
+
+-  All of the infrastructure surrounding the cuDF code, like the cluster setup,
+   diagnostics, JupyterLab environment, and so on, came for free, like any
+   other new Dask project.
+
+   Here is an image of my JupyterLab setup
+
+   <a href="https://matthewrocklin.com/blog/images/dask-cudf-environment.png">
+     <img src="https://matthewrocklin.com/blog/images/dask-cudf-environment.png"
+          alt="Dask + CUDA + cuDF JupyterLab environment"
+          width="70%">
+   </a>
+
+-  Our `df` object is actually just a normal Dask Dataframe.  We didn't have to
+   write new `__repr__`, `__add__`, or `.sum()` implementations, and probably
+   many functions we didn't think about work well today (though also many
+   don't).
+
+-  We're tightly integrated and more connected to other systems. For example if
+   we wanted to convert our dask-cudf-dataframe to a dask-pandas-dataframe then
+   we would use the standard cuDF function for this:
+
+   ```python
+   df = df.map_partitions(cudf.DataFrame.to_pandas)
+   ```
+
+   We don't have to write anything special like a separate `.to_dask_dataframe`
+   method nor handle other special cases.
+
+   The parallelism of Dask should be orthogonal to the choice of CPU/GPU.
+
+-  It's easy to switch hardware.  By avoiding separate `dask-cudf` code paths
+   it's easier to add cuDF to an existing Dask+Pandas codebase if we get a GPU,
+   or remove cuDF and use Pandas if we want our code to be runnable by people
+   who don't have GPUs.
+
+   We'll see more examples of this in the scaling section below.
+
+
+What's wrong with the example above
+-----------------------------------
+
+In general the answer is **many small things**.
+
+1.  The `cudf.read_csv` function doesn't yet support reading chunks from a
+    single CSV file, and so doesn't work well with very large CSV files.  We
+    had to split our large CSV files into many smalle CSV files first with
+    normal Dask+Pandas:
+
+    ```python
+    import dask.dataframe as dd
+    (df = dd.read_csv('few-large/*.csv')
+            .repartition(npartitions=100)
+            .to_csv('many-small/*.csv', index=False))
+    ```
+
+    (See [rapidsai/cudf #568](https://github.com/rapidsai/cudf/issues/568))
+
+2.  Many operations that used to work in dask-cudf like groupby-aggregations
+    and joins no longer work.  We're going to need to slightly modify many cuDF
+    APIs over the next couple months to more closely match their Pandas
+    equivalents.
+
+3.  I ran the timing cell twice because it currently takes a few seconds to
+    `import cudf` today.
+    [rapidsai/cudf #627](https://github.com/rapidsai/cudf/issues/627)
```

Comment:
Anything that uses Numba JIT may also have JIT overhead. And there's also always overhead the first time you run a CUDA kernel, so some things will always be faster the second time you run them.